faucet.go

```diff
@@ -725,11 +727,60 @@ func (l *lightningFaucet) invoicePage(w http.ResponseWriter, r *http.Request) {
 	http.Error(w, "", http.StatusMethodNotAllowed)
 }
 
+func (l *lightningFaucet) createUpdatesChans(id string) (
+	chan lnrpc.Invoice_InvoiceState, chan string, error) {
+	ctx := context.Background()
+
+	// Subscribe to receive updates from this invoice.
+	subInvoiceReq := &invoicesrpc.SubscribeSingleInvoiceRequest{
+		RHash: ongoingInvoices[id].RHash,
+	}
+	updatesStream, err := l.invoicesClient.SubscribeSingleInvoice(
+		ctx, subInvoiceReq)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create channels for streaming updates.
+	invoiceStateChan := make(chan lnrpc.Invoice_InvoiceState)
+	invoiceErrChan := make(chan string)
+
+	// Init a new routine to listen for updates
+	go func() {
+		for {
+			nextMsg, err := updatesStream.Recv()
+			if err != nil {
+				invoiceErrChan <- err.Error()
+				return
+			}
+			invoiceStateChan <- nextMsg.State
+		}
+	}()
+
+	return invoiceStateChan, invoiceErrChan, nil
+}
+
 func (l *lightningFaucet) handleWebsocket(w http.ResponseWriter, r *http.Request) {
 	id := mux.Vars(r)["id"]
 	//upgrader.CheckOrigin = func(r *http.Request) bool { return true }
-	invoiceStateChan := ongoingInvoices[id].invoiceStateChan
-	invoiceErrChan := ongoingInvoices[id].invoiceErrChan
+	//invoiceStateChan := ongoingInvoices[id].invoiceStateChan
+	//invoiceErrChan := ongoingInvoices[id].invoiceErrChan
+
+	invoiceStateChan, invoiceErrChan, err := l.createUpdatesChans(id)
```

Comment:
And pass in `r.Context()` here, so that the goroutine inside `createUpdatesChans` errors once the context for the request is closed.